src/components/spatial/SpatialSubscriber.js

```diff
@@ -2,210 +2,152 @@ import React, {
-      PubSub.unsubscribe(layerAddToken);
-      PubSub.unsubscribe(layerChangeToken);
-      PubSub.unsubscribe(layerRemoveToken);
-      PubSub.unsubscribe(cellsOnToken);
-      PubSub.unsubscribe(moleculesOnToken);
-      PubSub.unsubscribe(resetToken);
-      PubSub.unsubscribe(rasterAddToken);
-    };
-  }, [onReadyCallback, urls]);
-  const cellsCount = useMemo(() => (cells ? Object.keys(cells).length : 0), [cells]);
-  const [moleculesCount, locationsCount] = useMemo(() => {
-    if (!molecules) return [0, 0];
-    return [
-      Object.keys(molecules).length,
-      Object.values(molecules)
-        .map(l => l.length)
-        .reduce((a, b) => a + b, 0),
-    ];
-  }, [molecules]);
-  const updateStatus = useCallback(
-    message => PubSub.publish(STATUS_INFO, message),
-    [],
+    resetUrls();
+    resetReadyItems();
+    setAutoLayers([]);
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [loaders, dataset]);
+
+  // Get data from loaders using the data hooks.
+  const [cells, cellsCount] = useCellsData(
+    loaders, dataset, setItemIsReady, addUrl, false,
+    () => setAutoLayers(prev => ([...prev, DEFAULT_CELLS_LAYER])),
   );
-  const updateCellsSelection = useCallback(
-    selectedIds => PubSub.publish(CELLS_SELECTION, selectedIds),
-    [],
+  const [molecules, moleculesCount, locationsCount] = useMoleculesData(
+    loaders, dataset, setItemIsReady, addUrl, false,
+    () => setAutoLayers(prev => ([...prev, DEFAULT_MOLECULES_LAYER])),
   );
-  const updateCellsHover = useCallback(
-    hoverInfo => PubSub.publish(CELLS_HOVER, hoverInfo),
-    [],
+  // TODO: set up a neighborhoods default layer and add to autoLayers.
+  const [neighborhoods] = useNeighborhoodsData(
+    loaders, dataset, setItemIsReady, addUrl, false,
   );
+  const [cellSets] = useCellSetsData(
+    loaders, dataset, setItemIsReady, addUrl, false,
+  );
+  const [expressionMatrix] = useExpressionMatrixData(
+    loaders, dataset, setItemIsReady, addUrl, false,
+  );
+  // eslint-disable-next-line no-unused-vars
+  const [raster, imageLayerLoaders] = useRasterData(
+    loaders, dataset, setItemIsReady, addUrl, false,
+    (autoImageLayers) => {
+      setAutoLayers(prev => ([...prev, ...autoImageLayers]));
+    },
+  );
+
+  // Try to set up the layers array automatically if null or undefined.
+  useEffect(() => {
+    if (isReady && !layers) {
+      setLayers(sortLayers(autoLayers));
+    } else if (isReady && layers) {
+      // Layers were defined, but check whether channels for each layer were also defined.
+      // If channel / slider / domain definitions are missing, initialize in automatically.
+      initializeLayerChannelsIfMissing(layers, imageLayerLoaders)
+        .then(([newLayers, didInitialize]) => {
+          if (didInitialize) {
+            // Channels were only partially defined.
+            setLayers(newLayers);
+          }
+        });
+    }
+  }, [autoLayers, imageLayerLoaders, isReady, layers, setLayers]);
```

Comment:
Why does this need to happen?  How could `useRasterData` return incomplete data, or is that not what this is for?

Reply:
The `else if` part is used in order to auto-initialize the `.slider`, `.domain`, `.visible`, and/or `.color` properties of pre-defined layer and channel definitions. This way, a user can specify the channel selections they want to view in the view config, but they can still get auto-initialization for sliders, etc.

For example, I used this feature with the Vanderbilt IMS image pyramid aggregation algorithm comparison view config. I just put the layer and channel selections in the view config and let the auto initialization do the rest.

```js
"spatialLayers": {
      "ORIGINAL": [
        {
          "type": "raster",
          "index": 0,
          "opacity": 1,
          "colormap": "",
          "channels": [
            { "selection": { "channel": 0, "z": 0, "time": 0 } }, // notice the missing "sliders" "domain" and "color" properties
            { "selection": { "channel": 1, "z": 0, "time": 0 } },
          ]
        }
      ],
```

however if I know those properties ahead of time I can specify (some or all of) them and those properties will not be overridden

```js
"spatialLayers": {
      "ORIGINAL": [
        {
          "type": "raster",
          "index": 0,
          "opacity": 1,
          "colormap": "",
          "channels": [
            { "selection": { "channel": 0, "z": 0, "time": 0 }, "slider": [0, 100], "domain": [0, 200] },
            { "selection": { "channel": 1, "z": 0, "time": 0 }, "slider": [0, 100], "domain": [0, 200], "visible": false, "color": [255, 0, 0] },
          ]
        }
      ],
```

Reply:
Very cool!